tools/preprocess_pragmas.c

```diff
@@ -0,0 +1,146 @@
+
+// SPDX-FileCopyrightText: © 2024 ZeldaRET
+// SPDX-License-Identifier: CC0-1.0
+
+// Usage: preprocess_pragmas OOT_VERSION filename < source.c
+// The filename argument is only used for linemarkers.
+// Preprocess C source on stdin, writes to stdout
+// Replace `#pragma increment_block_number` with fake structs for controlling BSS ordering.
+// The names of these fake structs are expected to be increment_block_number_%d_%d with the first number indicating
+// the line number of the #pragma in the original source file. (this is for use by fix_bss.py)
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+const char str_pragma_increment_block_number[] = "#pragma increment_block_number";
+
+int main(int argc, char** argv) {
+#if 0
+    // for debugging with a specific file
+#undef stdin
+    FILE* stdin = fopen("src/code/fault.c", "r");
+    assert(stdin != NULL);
+#endif
+
+    if (argc != 3) {
+        fprintf(stderr, "Usage: preprocess_pragmas OOT_VERSION filename < source.c\n");
+        return EXIT_FAILURE;
+    }
+    char* const version = argv[1];
+    const int len_version = strlen(version);
+    char* const filename = argv[2];
+
+    char buf[32 * 1024];
+    char* const bufend = buf + sizeof(buf);
+    char* bufp = buf;
+    size_t nread;
+    int line_num = 1;
+    // whether the current line follows a #pragma increment_block_number,
+    // including continuation lines (lines after a \-ending line)
+    bool is_in_pragma = false;
+    // the line where the #pragma increment_block_number is
+    int pragma_line_number;
+    // how many fake structs to write to replace the current pragma
+    int n_fake_structs;
+
+    while ((nread = fread(bufp, 1, bufend - bufp, stdin)) != 0) {
+        bufp += nread;
+        char* last_line_return = NULL;
```

Comment:
I don't think "line return" is a thing, it's usually called "newline" (or "line feed" if it's being contrasted with "carriage return")